File: my_cpp_pkg/src/add_two_ints_client.cpp
#include <rclcpp/rclcpp.hpp>
#include <example_interfaces/srv/add_two_ints.hpp>

using namespace std::chrono_literals;
using namespace std::placeholders;

class AddTwoIntsClientNode : public rclcpp::Node
{
public:
  AddTwoIntsClientNode()
      : Node("add_two_ints_client")
  {
    pClient_ = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");

    RCLCPP_INFO(this->get_logger(), "Add Two Ints Client started.");
  }

  void callAddTwoInts(const int a, const int b)
  {
    auto pRequest = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
    pRequest->a = a;
    pRequest->b = b;

    while (!pClient_->wait_for_service(1s))
    {
      RCLCPP_WARN(this->get_logger(), "Waiting for Add Two Ints Server...");
    }

    pClient_->async_send_request(pRequest, std::bind(&AddTwoIntsClientNode::callbackCallAddTwoInts, this, _1));
  }

private:
  void callbackCallAddTwoInts(
      rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFutureWithRequest future_response)
  {
    auto pRequest = future_response.get().first;
    auto pResponse = future_response.get().second;

    RCLCPP_INFO(this->get_logger(), "%ld + %ld = %ld", pRequest->a, pRequest->b, pResponse->sum);
  }

  rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr pClient_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);

  auto pNode = std::make_shared<AddTwoIntsClientNode>();
  pNode->callAddTwoInts(4, 7);
  pNode->callAddTwoInts(156, 89);
  pNode->callAddTwoInts(1, -2);
  rclcpp::spin(pNode);

  rclcpp::shutdown();
  return 0;
}

File: my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial


class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")

        self.client_ = self.create_client(srv_type=AddTwoInts, srv_name="add_two_ints")

        self.get_logger().info("Add Two Ints Client started.")

    def call_add_two_ints(self, a, b):
        while not self.client_.wait_for_service(timeout_sec=1):
            self.get_logger().warn("Waiting for Add Two Ints Server...")

        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        future_response = self.client_.call_async(request)
        future_response.add_done_callback(
            partial(self.callback_call_add_two_ints, request=request)
        )

    def callback_call_add_two_ints(self, future_response, request):
        response = future_response.result()

        self.get_logger().info(
            str(request.a) + " + " + str(request.b) + " = " + str(response.sum)
        )


def main(args=None):
    rclpy.init(args=args)

    node = AddTwoIntsClientNode()
    node.call_add_two_ints(4, 7)
    node.call_add_two_ints(156, 89)
    node.call_add_two_ints(1, -2)
    rclpy.spin(node)

    rclpy.shutdown()


if __name__ == "__main__":
    main()
